### PR TITLE
Refactor Modals to Glimmer components

### PR DESCRIPTION
**Pull request by simonihmig**
Most components refactored to template-only.

### CODE REVIEW

File: addon/components/bs-modal-simple.hbs
Comment (simonihmig) [inline addon/components/bs-modal-simple.hbs:17]: Tests are failing due to the forwarding of arguments here. They do pass locally when `useDefaultValueIfUndefined`´ is set to true, as this is the same use case covered in #1248.

We could make it pass by wrapping every forwarding argument in a `{{bs-default}}`, requiring us to redefine all defaults here, which is pretty annoying. Also this brings up another issue: the `bs-default` helper actually does not follow the same semantics as the `@defaultValue` and `@arg` decorators, in that it will *always*  apply the default value if the argument is undefined. And there is no way around it, as it cannot distinguish between an argument not being used and and argument being passed with `undefined`. The only workaround - if we want to not risk any kind of breaking changes - would be to *not* use template-only components here and define defaults using `@arg`. 😟

This makes me think again if the approach in #1248 is really the best way forward, balancing the user's interests and ours as maintainers. Or if we shouldn't consider making the `useDefaultValueIfUndefined` behavior the default and only one. I think we still have a strong standpoint to not have to bump the major version, as eventually breaking code can be considered as relying on unintentional and undocumented behavior (e.g. I don't think there is anywhere in the API docs a reference to `undefined` as a legal value). Also I think there is a good chance that more or less no user is actually affected by this. So maybe we are making our lives way to hard here for no real (user's) benefit?

@jelhan thoughts?

Comment (jelhan) [reply addon/components/bs-modal-simple.hbs:17]: > This makes me think again if the approach in #1248 is really the best way forward, balancing the user's interests and ours as maintainers. Or if we shouldn't consider making the `useDefaultValueIfUndefined` behavior the default and only one.

I fully agree. I only introduced the `useDefaultValueIfUndefined` optional feature to unblock myself. There are strong use cases for it and I haven't come across any use case for the legacy behavior yet.

> I think we still have a strong standpoint to not have to bump the major version, as eventually breaking code can be considered as relying on unintentional and undocumented behavior (e.g. I don't think there is anywhere in the API docs a reference to `undefined` as a legal value). Also I think there is a good chance that more or less no user is actually affected by this. So maybe we are making our lives way to hard here for no real (user's) benefit?

I fully agree. As an alternative I would be also fine to have another major release. If it only includes these change and maybe drops `@triggerElement="parent-view"` upgrade should be easy for all applications. But we wouldn't compromise our strong SemVer guarantee.

Comment (simonihmig) [reply addon/components/bs-modal-simple.hbs:17]: So, seems we have agreement here then. I assume you are still on vacation, so I will create another PR to remove the optional feature again (in favor of making it the default). Sorry for the additional work this has caused! 😔

Comment (jelhan) [reply addon/components/bs-modal-simple.hbs:17]: Actually I'm back from vacation since today. But have a huge list of open topics - as always after vacation. 🙄 So would be awesome if you could take care of this one.

File: addon/helpers/bs-action-fallback.js
Comment (jelhan) [reply addon/helpers/bs-action-fallback.js:7]: This reminds me that I wanted to propose having `{{fn}}` helper to return a noop function if called without any argument.

Comment (simonihmig) [reply addon/helpers/bs-action-fallback.js:7]: Yeah, something to solve this upstream really makes sense IMO. Or `{{on ...}}` ignoring an `undefined` callback or something like that. You hereby have my support for writing an RFC! 😝

Comment (jelhan) [reply addon/helpers/bs-action-fallback.js:7]: It's discussed here: https://github.com/emberjs/rfcs/issues/660

File: addon/components/bs-modal-simple.hbs
Comment (simonihmig) [reply addon/components/bs-modal-simple.hbs:17]: I rebased this on top of master which now includes #1283, so that should hopefully pass tests.

File: addon/components/-bs-modal-simple.js
Comment (jelhan) [inline addon/components/-bs-modal-simple.js:57]:
```suggestion
      <Form.element @controlType="text" @label="first name" @property="firstname" autofocus={{true}} />
```

File: addon/components/-bs-modal-simple.js
Comment (jelhan) [inline addon/components/-bs-modal-simple.js:211]:
```suggestion
 * You can return `false` to prevent closing the modal automatically, and do that in your action by
```

File: addon/components/-bs-modal-simple.js
Comment (jelhan) [inline addon/components/-bs-modal-simple.js:212]:
```suggestion
 * setting `@open` to `false`.
```

File: addon/components/-bs-modal-simple.js
Comment (jelhan) [inline addon/components/-bs-modal-simple.js:231]:
```suggestion
 * size calculations and the like). This means that if `@fade={{true}}`, it will be shown in between the
```

**Review by jelhan — CHANGES_REQUESTED** [review]
Started a review some time ago but haven't had the time to finish it yet. Only reviewed API docs for `<BsModalSimple>` so far. I guess it's only a copy and paste. But noticed that some parts are outdated. Let's take the chance and fix that.

File: addon/helpers/bs-action-fallback.js
Comment (jelhan) [reply addon/helpers/bs-action-fallback.js:7]: I expect `{{or @actionHandler (fn)}}` to be the syntax supported by Ember out of the box in mid-term. Wondering if we should use a `{{bs-noop}}` together with `{{bs-default}}` instead of `{{bs-action-fallback}}` as an interim solution. Refactoring would be a task of search and replace in that case.

**Review by jelhan — CHANGES_REQUESTED** [review]
Reviewed the rest as well. Mostly general questions not strictly related to this refactoring. Keep request changes as I think the fixes to docs from last review should be included for sure.

File: addon/helpers/bs-action-fallback.js
Comment (jelhan) [reply addon/helpers/bs-action-fallback.js:7]: Published a RFC, which proposes to support an argumentless version of `{{fn}}` helper to create a noop function as we had discussed here: https://github.com/emberjs/rfcs/pull/676

File: addon/components/-bs-modal-simple.js
Comment (simonihmig) [reply addon/components/-bs-modal-simple.js:46]: For sure 🤦‍♂️

File: addon/components/bs-modal.js
Comment (simonihmig) [reply addon/components/bs-modal.js:690]: Oh, no, forgotten to clean up.

File: addon/components/bs-modal.js
Comment (simonihmig) [reply addon/components/bs-modal.js:713]: I guess for historic reasons, as this was `init()` previously. Will refactor.

File: addon/components/bs-modal.js
Comment (simonihmig) [reply addon/components/bs-modal.js:693]: Agree, changed.

File: addon/components/bs-modal.js
Comment (simonihmig) [reply addon/components/bs-modal.js:524]: Yes, but somehow not working. Like it didn't see the rendering changes when they callback ran. 🤔

File: addon/components/bs-modal.hbs
Comment (simonihmig) [reply addon/components/bs-modal.hbs:2]: How would you refactor that? 

> But I would prefer if changes to internal state does not trigger side effects as it makes reasoning about state and its flow much harder.

But in our case that needs to happen. When setting our inter `isOpen` to true, we need to create side effects that are not possible just by auto-updating our template, like changing the `<bod>` class, adding a `transitionEnd` listener etc.

File: addon/components/bs-modal/footer.hbs
Comment (simonihmig) [reply addon/components/bs-modal/footer.hbs:19]: Yeah, that works! without your syntax errors (no curlies) though :wink:

File: addon/components/bs-modal/header/-close.js
Comment (simonihmig) [reply addon/components/bs-modal/header/-close.js:5]: Idk. They are after all called template only Glimmer components, aren't they? Not that this matters much here, given this is a private component (only yielded), and extending is not supported anyways...

File: addon/components/bs-modal.js
Comment (jelhan) [reply addon/components/bs-modal.js:524]: It seems as if the callback itself also schedules something for `afterRender`: https://github.com/kaliber5/ember-bootstrap/blob/75f12adb7070fd0d931bc6f90d8941c5bb957425/addon/components/bs-modal.js#L433-L462 Maybe it's just not well specified what happens if scheduling something for the run loop you are currently?
:thinking: 

```js
schedule('afterRender', this, function() {
  schedule('afterRender', this, function() {
    // Same runloop or after render of next runloop? 😕
  });
});
```

File: addon/helpers/bs-action-fallback.js
Comment (simonihmig) [reply addon/helpers/bs-action-fallback.js:7]: > Wondering if we should use a {{bs-noop}} together with {{bs-default}} instead of {{bs-action-fallback}} as an interim solution

Makes sense, updated.

**Comment by simonihmig** [comment]
Thanks @jelhan for your super thorough review! Applied your suggestions.

The failing tests in beta and canary are due to https://github.com/buschtoens/ember-render-helpers/issues/282. 🤔

File: addon/components/bs-modal.hbs
Comment (jelhan) [reply addon/components/bs-modal.hbs]: > When setting our inter `isOpen` to true, we need to create side effects that are not possible just by auto-updating our template, like changing the `<bod>` class, adding a `transitionEnd` listener etc.

That shouldn't be caused by a side effect but explicitly executed at the same place as `isOpen` is changed.

> How would you refactor that?

Let me try to illustrate with some pseudo code.

```hbs
{{did-insert this.reactOnExternalVisibilityChanges}}
{{did-update this.reactOnExternalVisibilityChanges @open}}
```

```js
export default ModalComponent extends Component {
  @tracked isOpen = this.args.open;

  @action
  show() {
    if (this.isOpen) {
      // Modal is already open. Nothing to do.
    }

    this.isOpen = true;

    // ...
  }

  @action
  hide() {
    if (!this.isOpen) {
      // Modal is already hidden. Nothing to do.
    }

    this.isOpen = false;

    // ...
  }

  @action
  close() {
    if (this.args.onHide?.() !== false) {
      // Before this method calls `hide` method implicitly by setting
      // `this.isOpen = false`. This is the classical observer style.
      // Instead we should call the `hide` method explicitly.
      this.hide();
    }
  }

  @action  
  reactOnExternalVisibilityChanges() {
    if (this.args.open) {
      this.show();
    else {
      this.hide();
    }
  }
}
```

Not sure if I missed something. The general idea is that internal `isOpen` state is changed only through `show` and `hide` methods. Internal and yielded actions call these methods directly. External state passed in as `@open` argument is observed using `{{did-update}}` helper. If it changes `show` or `hide` method is executed.

`{{did-insert}}` is used caused DOM must be inserted before `show` method can be executed. I would have preferred to use `constructor` instead. But not sure if DOM is guaranteed to exist if scheduling if for `afterRender` or next runloop. That's why I didn't touched that part.

File: addon/components/bs-modal/footer.hbs
Comment (jelhan) [inline addon/components/bs-modal/footer.hbs:17]: Awesome that this pattern works! That means we can drop `_disabled` and similar arguments only kept for internal usage! :tada:

Comment (simonihmig) [reply addon/components/bs-modal/footer.hbs:17]: Not quite I think: that works only when we invoke (i.e. render) the component, but not when yielding it back. Which we do...

**Comment by simonihmig** [comment]
Updated `ember-render-helpers` which makes tests pass on CI.

@jelhan applied your suggestion regarding using `{{did-update ... @open}}` above (still need `this._isOpen` though) in the last commit. I did not have time to fiddle with the `schedule` timings you asked about though...

**Comment by jelhan** [comment]
@simonihmig Thanks a lot for pushing this over the finished line. Looks great!

**Pull request merged**
